Fix run on start

bugfood opened the pull request:
### What does this PR do?

1. Updates the default behavior of the `run_on_start` parameter to match the documentation and older salt behavior.
2. Update/extend unit tests accordingly.

### What issues does this PR fix or reference?
Fixes: #61964

### Initial Behavior
Jobs with seconds/minutes/hours/days parameters would run on startup by default.

### Regression from 391424c4e1859:
Such jobs no longer run on startup by default and need to be forced by `run_on_start: True'.

### New Behavior
Such jobs now run on startup by default again.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [X] Docs (N/A--this fixes the code to match the docs)
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

bugfood committed:
remove unused _next_scheduled_fire_time

This internal parameter is written to but never read from.

The equivalent internal parameter _next_fire_time was introduced in
391424c4e185; this parameter is actually used.

bugfood committed:
fix regression in running scheduled jobs at minion startup

According to the document:
https://docs.saltproject.io/salt/user-guide/en/latest/topics/scheduler.html

   By default, any job scheduled based on the startup time of the minion
   will run the scheduled job when the minion starts up.

This was broken by 391424c4e1859 and entrenched into the test suite by
various commits afterward.

Fix the code and adjust the unit tests to match. Add a new unit test to
ensure that disabling job run at startup still works.

bugfood committed:
remove redundant dry_run setting

This test is unlike the seconds, minutes, and hours tests; dry_run is
enabled unconditionally.

bugfood committed:
fix typos in test docstrings

OrangeDog commented:
They should still, however, run at a random interval after startup depending on the `splay` parameter. Is that still the case?

Tests like `test_eval_splay` probably need some adjustment.

bugfood committed:
make splay testing more precise

1. Ensure that jobs are run at minion startup as per the default
   `run_on_start: True`.
2. Ensure that jobs are not run before the splayed interval.
3. Update _check_last_run docstring to match behavior.

bugfood commented:
I do think that the `splay` parameter is still respected appropriately. I added a commit, just now, to make the splay testing more precise to verify this.

bugfood commented:
I think the test failure on this PR is unrelated to the changes I made.

bugfood commented:
Thanks for the interest in this old PR. I have not been active with saltstack work in the interim, but of course you can still use the changes if they are any good.

Wouldn't it be better to rebase this branch against current master rather than merge master into this branch?

twangboy commented:
Please rebase and fix the failing pre-commit

bugfood commented:
> Please rebase and fix the failing pre-commit

Rebased and amended.

The pre-commit complaints seem to be about:
* Wanting a rename of the file in `changelog/`; done.
* Wanting formatting via Black; done.

I think the CI pipeline needs to triggered to run.

bugfood commented:
> @bugfood
> 
> Is this regression present in `3006.x`?

I think so, from looking at the code--there don't seem to have been any significant changes to the `eval` function on the `v3006.13` branch in the time since I originally made this PR.

Thank you for the merge.

-Corey
